### ISSUE TITLE
Unused method removal removes methods passed to cognew

### ISSUE THREAD

**Issue by bweir** [parallaxinc/OpenSpin#17]
`MusicPlayer` is a private function passed to a cognew, but is used nowhere else, so `-u` removes the function.

Compiling without unused method removal:
```
$ ./openspin ~/Projects/lamestation-sdk/games/Brettris/Brettris.spin -L ~/Projects/lamestation-sdk/
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.72 Compiled on Mar 19 2015 21:49:36
Compiling...
/home/brett/Projects/lamestation-sdk/games/Brettris/Brettris.spin
|-LameLCD.spin
  |-Pinout.spin
|-LameGFX.spin
|-LameAudio.spin
  |-Pinout.spin
|-LameMusic.spin
  |-LameAudio.spin
    |-Pinout.spin
|-LameControl.spin
  |-Pinout.spin
  |-LameFunctions.spin
|-LameFunctions.spin
|-gfx_bretrominos.spin
|-gfx_wall.spin
|-gfx_font4x6_w.spin
|-gfx_scorebox.spin
|-song_tetris.spin
|-gfx_bretris_logo.spin
|-gfx_gameover.spin
Done.
Program size is 11752 bytes
```
With unused method removal enabled:
```
$ ./openspin ~/Projects/lamestation-sdk/games/Brettris/Brettris.spin -L ~/Projects/lamestation-sdk/ -u
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.72 Compiled on Mar 19 2015 21:53:47
Compiling...
/home/brett/Projects/lamestation-sdk/games/Brettris/Brettris.spin
|-LameLCD.spin
  |-Pinout.spin
|-LameGFX.spin
|-LameAudio.spin
  |-Pinout.spin
|-LameMusic.spin
  |-LameAudio.spin
    |-Pinout.spin
|-LameControl.spin
  |-Pinout.spin
  |-LameFunctions.spin
|-LameFunctions.spin
|-gfx_bretrominos.spin
|-gfx_wall.spin
|-gfx_font4x6_w.spin
|-gfx_scorebox.spin
|-song_tetris.spin
|-gfx_bretris_logo.spin
|-gfx_gameover.spin
Done.
NOT IMPLEMENTED
NOT IMPLEMENTED
NOT IMPLEMENTED
NOT IMPLEMENTED
|-LameLCD.spin
|-LameGFX.spin
|-LameAudio.spin
|-LameMusic.spin
  |-LameAudio.spin
LameMusic.spin(48:12) : error : Expected an expression term
Line:
    cognew(MusicPlayer, @MusicPlayerStack)
Offending Item: MusicPlayer
```
I'm building off a fresh checkout of parallaxinc/OpenSpin.

I did try a few projects that didn't use this feature and it reduced the binary size significantly, so it's looking good so far. :)

**Comment by reltham** [comment]
Calling this fixed now, but please re-open if you find problems when testing.

**Comment by bweir** [comment]
Hi there,

Thanks for the update. It looks like the issue with cognew is fixed. However, now it really seems to dislike the `Sleep` command used in my `LameFunctions` library.

Here is the build with unused function removal disabled:

```
$ ./openspin ~/Projects/lamestation-sdk/games/Brettris/Brettris.spin -L ~/Projects/lamestation-sdk/
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.74 Compiled on May 11 2015 02:34:07
Compiling...
/home/brett/Projects/lamestation-sdk/games/Brettris/Brettris.spin
|-LameLCD.spin
  |-Pinout.spin
|-LameGFX.spin
|-LameAudio.spin
  |-Pinout.spin
|-LameMusic.spin
  |-LameAudio.spin
    |-Pinout.spin
|-LameControl.spin
  |-Pinout.spin
  |-LameFunctions.spin
|-LameFunctions.spin
|-gfx_bretrominos.spin
|-gfx_wall.spin
|-gfx_font4x6_w.spin
|-gfx_scorebox.spin
|-song_tetris.spin
|-gfx_bretris_logo.spin
|-gfx_gameover.spin
Done.
Program size is 11752 bytes
```

And when enabled:

```
$ ./openspin ~/Projects/lamestation-sdk/games/Brettris/Brettris.spin -L ~/Projects/lamestation-sdk/ -u
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.74 Compiled on May 11 2015 02:34:07
Compiling...
/home/brett/Projects/lamestation-sdk/games/Brettris/Brettris.spin
|-LameLCD.spin
  |-Pinout.spin
|-LameGFX.spin
|-LameAudio.spin
  |-Pinout.spin
|-LameMusic.spin
  |-LameAudio.spin
    |-Pinout.spin
|-LameControl.spin
  |-Pinout.spin
  |-LameFunctions.spin
|-LameFunctions.spin
|-gfx_bretrominos.spin
|-gfx_wall.spin
|-gfx_font4x6_w.spin
|-gfx_scorebox.spin
|-song_tetris.spin
|-gfx_bretris_logo.spin
|-gfx_gameover.spin
Done.
|-LameLCD.spin
|-LameGFX.spin
|-LameAudio.spin
|-LameMusic.spin
  |-LameAudio.spin
|-LameControl.spin
LameControl.spin(77:12) : error : Expected a subroutine name
Line:
        fn.Sleep(20)
Offending Item: Sleep
```

**Comment by reltham** [comment]
Can you point me at where I can get this code to test compiling under the debugger?

**Comment by bweir** [comment]
Sure, here's a distribution of Pikemanz. Thanks.
http://lamestation.net/downloads/misc/spin/Pikemanz.zip

**Comment by reltham** [comment]
Okay, thanks for the link to the project. I found the problem and submitted a fix, version 1.00.75.
I was able to successfully compile Pikemanz, but I can't test it for actually working.

**Comment by bweir** [comment]
Ok, so I tried it out and it seems to be running without any issues, and it's definitely reducing the size of the resulting binary, which is pretty cool. =P

```
17488	/home/brett/Projects/lamestation-sdk/games/Pikemanz/00_Pikemanz.binary
18364	/home/brett/Projects/lamestation-sdk/games/Pikemanz/00_Pikemanz.binary
```

So since I was able to compile Pikemanz, I decided to try some more extensive testing, and there are still some issues with this new feature.

If you download the vanilla Spin library, there is a script called test.sh that will allow you to compile every single project as is. Just define SPINC as `openspin -u` or drop the `openspin` binary into the root of the project. Get it here:

```
git clone https://github.com/parallaxinc/spin-standard-library
```

Compiling the project with `test.sh`:

```
$ ./test.sh 

./openspin -u -L . ./demos/PropellerRTC_Emulator_DEMO.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/PropellerRTC_Emulator_DEMO.spin
|-PropellerRTC_Emulator.spin
|-FullDuplexSerial.spin
Done.
|-PropellerRTC_Emulator.spin
PropellerRTC_Emulator.spin(17:12) : error : Expected an expression term
Line:
    cognew(Run(TimeAddress),@Stack)
Offending Item: Run


./openspin -u -L . ./demos/SingingDemo.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/SingingDemo.spin
|-VocalTract.spin
|-StereoSpatializer.spin
Done.
|-VocalTract.spin
./demos/SingingDemo.spin(73:10) : error : Expected a subroutine name
Line:
  stereo.start(@input, @buffer, buffer_size, 11, -1, 10, -1) 'start spatializer
Offending Item: start


./openspin -u -L . ./demos/Stack Length Demo.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/Stack Length Demo.spin
|-Stack Length.spin
  |-Parallax Serial Terminal.spin
Done.
|-Stack Length.spin
  |-Parallax Serial Terminal.spin
./demos/Stack Length Demo.spin(56:10) : error : Expected an expression term
Line:
  cognew(Toggle(Pin, DelayMS, Count), @Stack)
Offending Item: Toggle


./openspin -u -L . ./demos/SingingDemoSeven.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/SingingDemoSeven.spin
|-VocalTract.spin
|-StereoSpatializer.spin
Done.
|-VocalTract.spin
./demos/SingingDemoSeven.spin(72:10) : error : Expected a subroutine name
Line:
  stereo.start(@input, @buffer, buffer_size, 11, -1, 10, -1) 'start spatializer
Offending Item: start


./openspin -u -L . ./demos/Servo32v7_RampDemo.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/Servo32v7_RampDemo.spin
|-Servo32v7.spin
  |-Servo32_Ramp_v2.spin
Done.
./demos/Servo32v7_RampDemo.spin(23:11) : error : Expected a subroutine name
Line:
    SERVO.Start                 'Start Servo handler
Offending Item: Start


./openspin -u -L . ./demos/RCTIME_background_DEMO.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/RCTIME_background_DEMO.spin
|-TV.spin
|-Graphics.spin
|-RCTIME.spin
|-Numbers.spin
Done.
|-TV.spin
|-Graphics.spin
|-RCTIME.spin
RCTIME.spin(27:27) : error : Expected an expression term
Line:
  cogon := (cog := cognew(RCTIME(Pin,State,RCValueAddress),@RCStack)) > 0
Offending Item: RCTIME

./test.sh: line 24: 14726 Segmentation fault      (core dumped) ${SPINC} -L . "$line" > /dev/null

./openspin -u -L . ./demos/VGA_Text_Demo.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/VGA_Text_Demo.spin
|-VGA_Text.spin
  |-VGA.spin
Done.
./test.sh: line 24: 14728 Segmentation fault      (core dumped) ${SPINC} -L . "$line"


./openspin -u -L . ./demos/COIL_demo.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/COIL_demo.spin
|-TV.spin
|-Graphics.spin
|-COILREAD.spin
  |-Clock.spin
Done.
|-TV.spin
|-Graphics.spin
|-COILREAD.spin
COILREAD.spin(34:27) : error : Expected an expression term
Line:
  cogon := (cog := cognew(COILREAD(PingPin,CoilPin,ReadPin,COILAddress),@COILStack)) > 0
Offending Item: COILREAD

./test.sh: line 24: 14732 Segmentation fault      (core dumped) ${SPINC} -L . "$line" > /dev/null

./openspin -u -L . ./demos/TV_Text_Demo.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/TV_Text_Demo.spin
|-TV_Text.spin
  |-TV.spin
Done.
./test.sh: line 24: 14734 Segmentation fault      (core dumped) ${SPINC} -L . "$line"


./openspin -u -L . ./RCTIME.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./RCTIME.spin
Done.
./RCTIME.spin(27:27) : error : Expected an expression term
Line:
  cogon := (cog := cognew(RCTIME(Pin,State,RCValueAddress),@RCStack)) > 0
Offending Item: RCTIME


./openspin -u -L . ./PropellerRTC_Emulator.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./PropellerRTC_Emulator.spin
Done.
./PropellerRTC_Emulator.spin(17:12) : error : Expected an expression term
Line:
    cognew(Run(TimeAddress),@Stack)
Offending Item: Run


./openspin -u -L . ./COILREAD.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./COILREAD.spin
|-Clock.spin
Done.
./COILREAD.spin(34:27) : error : Expected an expression term
Line:
  cogon := (cog := cognew(COILREAD(PingPin,CoilPin,ReadPin,COILAddress),@COILStack)) > 0
Offending Item: COILREAD
```

There are also a couple remaining issues in the LameStation SDK. You can download it here (there are lots of submodules so use `--recursive`).

```
git clone --recursive https://github.com/lamestation/lamestation-sdk.git
```

And using the same `test.sh` script.

```
$ SPINC='../openspin -u' ./test.sh 

../openspin -u -L . ./demos/audio/PlayMIDI.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./demos/audio/PlayMIDI.spin
|-LameSerial.spin
|-LameAudio.spin
  |-Pinout.spin
|-LameGFX.spin
|-LameLCD.spin
  |-Pinout.spin
Done.
|-LameSerial.spin
./demos/audio/PlayMIDI.spin(90:11) : error : Expected a subroutine name
Line:
    audio.Start
Offending Item: Start


../openspin -u -L . ./games/piXel/piXel_Effects.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.75 Compiled on May 13 2015 15:09:44
Compiling...
./games/piXel/piXel_Effects.spin
|-LameGFX.spin
|-piXel_Sound.spin
  |-LameAudio.spin
    |-Pinout.spin
  |-LameFunctions.spin
|-gfx_boom.spin
Done.
|-piXel_Sound.spin
piXel_Sound.spin(85:1) : error : No PUB routines found
Line:
End Of File
Offending Item: N/A
```

Thanks!

**Comment by reltham** [comment]
Ok, it looks like there is still an issue with the indexing.  I'll try to get this fixed ASAP.

**Comment by bweir** [comment]
Thanks, I'm looking forward to it!

**Comment by reltham** [comment]
Ok, I've submitted 1.00.76 with several fixes.
I was able to successfully compile everything in the spin standard library as well as all the lamestation sdk demos.
Please check it out and make sure the compiled results for lamestation stuff runs properly.

**Comment by bweir** [comment]
Hi Roy,

Thanks for your help. Sorry I haven't replied in a few days. I've been away from my computer doing other things.

1.00.76 is looking pretty good. As of now, there are three files that aren't compiling. The rest compile, and I've been spot-checking games and they seem to work alright.

In the LameStation SDK:
```
../openspin -u -L . ./games/piXel/piXel_Effects.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.76 Compiled on May 27 2015 03:53:03
Compiling...
./games/piXel/piXel_Effects.spin
|-LameGFX.spin
|-piXel_Sound.spin
  |-LameAudio.spin
    |-Pinout.spin
  |-LameFunctions.spin
|-gfx_boom.spin
Done.
|-piXel_Sound.spin
piXel_Sound.spin(85:1) : error : No PUB routines found
Line:
End Of File
Offending Item: N/A
```

And in the Spin library:

```
./openspin -u -L . ./demos/SingingDemo.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.76 Compiled on May 27 2015 03:53:03
Compiling...
./demos/SingingDemo.spin
|-VocalTract.spin
|-StereoSpatializer.spin
Done.
|-VocalTract.spin
./demos/SingingDemo.spin(73:10) : error : Expected a subroutine name
Line:
  stereo.start(@input, @buffer, buffer_size, 11, -1, 10, -1) 'start spatializer
Offending Item: start


./openspin -u -L . ./demos/SingingDemoSeven.spin
Propeller Spin/PASM Compiler 'OpenSpin' (c)2012-2015 Parallax Inc. DBA Parallax Semiconductor.
Version 1.00.76 Compiled on May 27 2015 03:53:03
Compiling...
./demos/SingingDemoSeven.spin
|-VocalTract.spin
|-StereoSpatializer.spin
Done.
|-VocalTract.spin
./demos/SingingDemoSeven.spin(72:10) : error : Expected a subroutine name
Line:
  stereo.start(@input, @buffer, buffer_size, 11, -1, 10, -1) 'start spatializer
Offending Item: start
```

I'm not sure why they would only be having an issue on my machine. Would there be any way I can debug it in the way that you are, but under Linux? Perhaps there are some platform differences.

**Comment by reltham** [comment]
First off, when I downloaded the lamestation sdk zip, it didn't include the source files for anything in the games folder. It had everything in the demos folder, and that's what I had been testing against. So I never tested the piXel game compiling because it wasn't in the zip I got from github. Odd.

When I retested the spin standard library stuff I did get the errors with the Singing demos.

What happened was that I had made one change to handle arrays of objects (like what is used in the Singing demos), but then later made a change to handle multiple instances of the same object being used under different names.  When I tested things for the last change I had only tested the things that failed with that specific problem before. However, these two changes interfered with each other, and in fact I only needed the second change and removed the first one. Now everything in the spin library compiles, and I bet piXel will compile too.

Sorry about failing to properly retest everything after all the changes.
